### PR TITLE
New Group Page Reformulated

### LINKED ISSUE #1336: [Consumer UI Refresh] Spruce up the New Group page
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

We implemented a new page at `#/new/group` that is not yet accessible by clicking around in the UI. In this view, multiple users can be selected and a chat can be started with them. 

**Deliverables:**
- We must update the UI for showing the "selected" state by using the icon asset provided
- Make various design adjustments based on feedback of our design team 
- The page must also be moved into a modal that is type `right_docked` (similar to the `Settings` view)

new group

### PR DESCRIPTION
**Pull request by lucas-neuhaus-dev**
@marcaaron @shawnborton 

### Details
Changed the `SettingsModal` to function as a generic and reusable `RightDockedModal`, which takes additional `title` and `Page` props now.
The `NewGroupPage` now uses this new `RightDockedModal` to show up.
Added the Checkmark Icon on selected contacts on `OptionRow`.
Added a hover style for the `Create Group` button on the `NewGroupPage`.

#### I mentioned I would try to add, but did not
On the issue page, I mentioned I would try to add the hover style for the `OptionRow`s. I found some not easy to deal with regressions within the `ChatSwitcherView`, and because that will eventually be redone on issue #1338, I believe it is easier to deal with the hover on that issue.
The background when selecting via arrow keys is full width now, at least.

#### IMPORTANT - Bugs found
 I found the following bugs, and after trying to deal with them without much success, I talked to Marc. He told me to create the PR with what I have for now, and then we decide if we try to fix them now or on the future.

1. Wrong top spacing with `RightDockedModal` on some Android devices.
This can be seen in the Android screenshot below. The cause of it, after some testing, is that the modal does not go really full screen on some Android devices, it doesn't go over the top status bar. Because of that, when we add the `topSafeAreaPadding`, it actually ends up adding this value twice.
I found some stuff on that I will try later [link 1](https://github.com/react-native-modal/react-native-modal/issues/442), [link 2](https://stackoverflow.com/questions/39126194/react-native-modal-component-not-covering-entire-screen-on-android-when-using);
2. For testing on mobile, I went ahead and added the `redirect(ROUTES.NEW_GROUP)` on the `CreateMenu` button, as it will be in the future (issue #1337), for some reason it did not work on my iPhone emulator. To actually be able to test it on the iPhone, I had to temporarily change the redirect route of the `navigateToSettings()` function on the `HomePage.js` file. The redirect worked on all other platforms as intended.
3. On the iPhone emulator, when the keyboard was active on the `NewGroupPage`, the `Create Group` button did not show up. I found this extremely weird and don't know if there is just something wrong with my emulator. The button showed up when I dismissed the keyboard.

### Fixed Issues
Fixes #1336 

#### Web, Mobile web
1. Add the `/new/group` route on the URL, check if the `NewGroupPage` modal shows up and everything works.

#### Android, iOS, Desktop
Because the actual redirect button action is not meant to be added on this PR, it needs some manual changes for testing. 
1. The easier way is to change the `navigateToSettings()` function on `HomePage.js`, change the route to `ROUTE.NEW_GROUP`;
2. Click the user avatar;
3. Check if the `NewGroupPage` shows up and everything works.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
web

#### Mobile Web
web mobile

#### Desktop
desktop

#### iOS
ios

#### Android
android

### CODE REVIEW

File: src/pages/NewGroupPage.js
Comment (shawnborton) [inline src/pages/NewGroupPage.js:253]: Can we try this with no padding on the top?

File: src/styles/themes/default.js
Comment (shawnborton) [inline src/styles/themes/default.js:13]: I don't anticipate having multiple greens the same way we have a spectrum of grays. So I wonder if it makes more semantic sense to just keep `colors.green` and then use `colors.greenHover` or something like that?

File: src/components/OptionsSelector.js
Comment (shawnborton) [inline src/components/OptionsSelector.js:157]: Could we do `pb3` here so that the vertical padding is the same?

**Comment by shawnborton** [comment]
Looks great! Left a few minor comments, and I have a few more but I am curious if @marcaaron thinks they should be addressed in this PR.

The message for too many participants needs some styling love:
image

It should have a padding around it of `p5` and the heading needs to be in the proper heading color (currently it is black, which must mean no color is being passed to it)

Other notes:
image

- The name/email should be in our GTA Bold style/weight
- We should not show the pin icons here
- The checkmark circle should use our `themeColors.border` color

**Comment by shawnborton** [comment]
Also separate to this PR - after using this, it makes me wonder how useful the arrow keys even are for this particular flow. I think the arrow keys make sense for the "Search" flow where you want to quickly switch between chats, but for these New flows I can see where they don't add much and I would opt to not have that functionality here.

**Comment by lucas-neuhaus-dev** [comment]
@shawnborton Working on the review/requests.

Just a note on the checkmark circle color. I changed the original `themeColors.border` because I thought it looked weird when the `OptionRow` was selected (via arrow keys). It is the same as the background. It can be seen in the image below.

I will change it back on the next commit with the fixes.

circle color

**Comment by marcaaron** [comment]
>Wrong top spacing with RightDockedModal on some Android devices

I think we can address this bug in a follow up issue as it already existed. Best thing to do would be to create a new issue for this and then we would give you the first option of fixing it (if you like) outside of this PR.

>For testing on mobile, I went ahead and added the redirect(ROUTES.NEW_GROUP) on the CreateMenu button, as it will be in the future

I think it's fine to go ahead and hook this up now to facilitate testing if we need to we can remove the call to `ChatSwitcher.show()`

>On the iPhone emulator, when the keyboard was active on the NewGroupPage, the Create Group button did not show up. I found this extremely weird and don't know if there is just something wrong with my emulator. The button showed up when I dismissed the keyboard.

I'm seeing this too and I think it might have something to do with the fact that we are trying to present two modals at once. e.g. the `CreateMenu` itself is a `Modal` and must be completely closed before we open the `NewGroupPage` in the `RightDockedModal`. I think we can prove there is a race condition here by adding a `setTimeout` to our redirect call.

```js
        {
            icon: Users,
            text: 'New Group',
            onPress: () => {
                props.onItemSelected();
                setTimeout(() => {
                    redirect(ROUTES.NEW_GROUP);
                }, 500);
            },
        },
```

But this is not really ideal and just to provide a hint at what could be happening 🤔 

Seems like this is a known issue
https://github.com/react-native-modal/react-native-modal/issues/30

And there is also a note here in the docs about how to handle this -> 
https://github.com/react-native-modal/react-native-modal#i-cant-show-multiple-modals-one-after-another

**Comment by shawnborton** [comment]
> Just a note on the checkmark circle color. I changed the original themeColors.border because I thought it looked weird when the OptionRow was selected (via arrow keys). It is the same as the background. It can be seen in the image below.

Ahh, that makes sense. So a couple of workarounds for now (assuming we keep the arrow behavior):
- Make the highlighted row BG color use `gray1` instead of `gray2`
- Detect if a row is highlighted, and if so, make the border color of the circle a little darker
- Or the easiest thing to do is probably just give the checkmark circle a white background (`themeColors.componentBG`)

**Comment by lucas-neuhaus-dev** [comment]
@shawnborton Fixed your requests (hopefully did not forget anything)
For hiding the Pin icon, I ended up with quite a bit of prop drilling, with a new prop I called `hideOptionIcons`, but I believe that is the way to go.
The prop defaults to `false`, so it does not cause any regression elsewhere.

The three options you gave, I ended up choosing on changing the BG to `gray1` (`themeColor.buttonHoveredBG`). But I tried the other ones and will show below if you want to check them out.


option 1
option 1
option 2
option 2
option 3
option 1

@marcaaron 
> I think it's fine to go ahead and hook this up now to facilitate testing if we need to we can remove the call to ChatSwitcher.show()

Will do

> I'm seeing this too and I think it might have something to do with the fact that we are trying to present two modals at once

I would never have thought of that. Will take a look at the links you provided (and somewhere else with necessary), thanks!

File: src/pages/NewGroupPage.js
Comment (lucas-neuhaus-dev) [reply src/pages/NewGroupPage.js]: Done!

File: src/components/OptionsSelector.js
Comment (lucas-neuhaus-dev) [reply src/components/OptionsSelector.js]: Done!

File: src/styles/themes/default.js
Comment (lucas-neuhaus-dev) [reply src/styles/themes/default.js]: Done!

**Comment by shawnborton** [comment]
I like Option 1 if we can do it. One minor comment though - can we actually make the background color of the circle (not the border color) use `themeColors.componentBG`?

> Is it just changing the font-family to GTA_BOLD? I was a little confused. Especially because it looks like the value of GTA and GTA_BOLD is the same on the styles/fontFamily file.

The font family needs to be `GTA_BOLD` and the font weight also needs to be bold, so basically:

```
fontFamily: fontFamily.GTA_BOLD,
fontWeight: fontWeightBold,
```

However, maybe you can just add `styles.textStrong` to the display name since that style already exists?

**Comment by lucas-neuhaus-dev** [comment]
@shawnborton Got option 1 with the `themeColors.componentBG` background.

About the bold, I didn't like very much both the name and the email bold and went to look at the Figma file you sent me some days ago.
There, I found examples where the name is bold, and the email is normal, but with `gray5`. So that is what I did. 
If you want me to change back the email to `gray4` and bold, just let me know.

_What I did_
what i did

_Name and email bold_
both bold

The text style had to be done conditionally to not break the `ChatSwitcherView`, because of that I decided to change the prop I created before _(`hideOptionIcons`)_ to a new one called `isOnRightDockedModal`, and use it for both the text styling and the icon hiding.

**Comment by shawnborton** [comment]
Ah, sorry for any confusion - you are correct that this is the intended style:
image


Can you also make sure the display name (the bold part) uses `themeColors.heading` too? 

Pulling your branch now to check out the other changes. Will report back with any feedback I have!

**Comment by shawnborton** [comment]
For the message that is displayed when there is too many participants, I think we can actually remove the top padding (since the search input above it has bottom padding):

Current:
image


Proposed:
image

Also, the background highlight color feels strange in that the arrow keys only work if you are focused into the search input. Also, if you are focused into the search and you use the arrow keys and then press enter to select someone, I would expect the view to scroll back to the top. Anyways, I wonder if we could just remove this highlight unless you are focused into the search, or just remove that functionality here altogether?

**Comment by shawnborton** [comment]
Last small comment.... could we add `ml2` to the checkboxes so that there is a little bit of space between the truncated text and the circle? 

Current:
image


Proposed:
image

**Comment by lucas-neuhaus-dev** [comment]
> And the email part (the small font) should use themeColors.textSupporting

That would mean the email style would be exactly the same as before (as it is on the `ChatSwitcherView`). Removed the conditional styling.
I just realized there is a `textUnreadStyle` on the `OptionRow` file and is definitely what you want. I really missed it before.

Added the left margin on the checkbox;
Removed top padding on the header message;
Removed the keyboard functionality on `RightDockedModal`s.

New style:
new style

**Comment by shawnborton** [comment]
Thanks, looks good to me!

**Review by shawnborton — DISMISSED** [review]

**Comment by shawnborton** [comment]
Actually just tested this on iOS and I found the scrolling to be a bit buggy/laggy, and I ended up finding myself in this situation:
image

**Comment by shawnborton** [comment]
And one more small comment - can we make the small `sectionHeader` text use the color `themeColors.heading` as well please? Thanks!

**Comment by lucas-neuhaus-dev** [comment]
@shawnborton Noticed this problem on iOS as well. I am trying to find out the reason, I'm still clueless.

**Comment by lucas-neuhaus-dev** [comment]
@marcaaron I think I am going to try to transform the `<SectionList />` on the `OptionList` file in a `<FlatList />`. 
I could not find much about the iOS laggy and broken header stuff, maybe this could do it.
And it might be a good optimization for the list since it can have a ton of rows.

**Comment by marcaaron** [comment]
>transform the <SectionList /> on the OptionList file in a <FlatList />

@lucas-neuhaus-dev hmm I think that might be a lot of work and would probably advise against it unless there is some clear reason to do this? lagginess could have a lot of causes also the app is running in development mode most likely?

**Comment by marcaaron** [comment]
I would suggest we focus on getting this done and then look at performance improvements and other small bugs in another issue.

**Comment by lucas-neuhaus-dev** [comment]
Ok! That would indeed be a lot of work.
I believe I got all of Shawn's reviews done, and bug's 2 and 3 described on the PR fixed. But it probably still needs some code cleaning.
I am going to take a look at everything again and wait for your reviews.

**Comment by marcaaron** [comment]
Testing iOS now.. it's not laggy for me, but the list doesn't seem to scroll so something is probably wrong with the styling?

**Comment by marcaaron** [comment]
@shawnborton One idea for a future improvement (not in this PR) is to show some kind of loading indicator for the "Create Group" button. It can take a few seconds to create the chat and navigate the user.

**Comment by lucas-neuhaus-dev** [comment]
@marcaaron Got it.
SectionList has this `stickySectionHeadersEnabled` enabled by default on iOS, that what was causing the headers to go above the rows on the list. [ref](https://reactnative.dev/docs/sectionlist#stickysectionheadersenabled)
The scrolling problem: The `swipeDirection` of the modal component was interfering with the gesture capture. I tested with `swipeDirection = undefined` and worked beautifully. I hope it is not a problem to remove it.

Commit coming soon.

**Comment by marcaaron** [comment]
Amazing! I think it's fine. There is a chance we would change this in the future to use `react-navigation` so we can live without the swipe gesture for now. Thanks for looking into that.

File: src/components/OptionsList.js
Comment (marcaaron) [inline src/components/OptionsList.js:41]: It might make more sense to name this for the behavior we want to allow or prevent (which is focusing options with arrow keys) rather than where it will be used.

What if instead this was called `canFocusOptions`?

File: src/components/RightDockedModal.js
Comment (marcaaron) [inline src/components/RightDockedModal.js:47]: Since we are rendering this as a child maybe instead of specifying a `Page` prop we can use `this.props.children` allowing this component to render whatever it likes?

File: src/pages/NewGroupPage.js
Comment (marcaaron) [inline src/pages/NewGroupPage.js:262]: `{backgroundColor: themeColors.buttonSuccessHoveredBG}` should be a style if possible. 
Since we are trying to avoid having inline styles everywhere for consistency.

File: src/pages/home/sidebar/OptionRow.js
Comment (marcaaron) [inline src/pages/home/sidebar/OptionRow.js:64]: This maybe should be `showOptionsAsUnread` with a default value of `true`

Comment (marcaaron) [reply src/pages/home/sidebar/OptionRow.js:64]: Or maybe just a prop to `showAdditionalOptionStates` which would cover unread, draft comment, pinned etc

File: src/styles/getModalStyles.js
Comment (marcaaron) [inline src/styles/getModalStyles.js:123]: Maybe let's leave a comment here as a hint in case someone decides they want a gesture and then breaks everything?

**Review by marcaaron — CHANGES_REQUESTED** [review]
Testing well. Nicely done! I do have some requests. Thanks!

**Comment by marcaaron** [comment]
Also, not too sure why but I am no longer being redirected to a group chat after it has been created - will take a look. 

Other than that testing perfectly.

One last thing, there is some code that I think is being duplicated

https://github.com/Expensify/Expensify.cash/blob/c05d8523f2639966acacbc3b2665c530c836fe12/src/pages/NewGroupPage.js#L211-L225

If you look there the styles were intended to be removed. But now we have two headers it's just that one is hidden.

**Comment by marcaaron** [comment]
Ah yes, seems the sidebar is still shown once we redirect to the report, but only can notice this on small width screens.

File: src/components/CreateMenu.js
Comment (lucas-neuhaus-dev) [reply src/components/CreateMenu.js:42]: It makes a lot of sense to not use state here, just a reference. We definitely don't need a re-render.
Just to make sure, is it what you meant?
```javascript
(...)
this.setOnModalHide = this.setOnModalHide.bind(this);
this.resetOnModalHide = this.resetOnModalHide.bind(this);
this.onModalHide = () => {};
(...)
setOnModalHide() {
    this.onModalHide = () => redirect(ROUTES.NEW_GROUP);
}
(...)
```

Comment (lucas-neuhaus-dev) [reply src/components/CreateMenu.js:42]: The above, plus adjusting everything related to the changes.

Comment (lucas-neuhaus-dev) [reply src/components/CreateMenu.js:42]: I was thinking of the `setOnModalHide` to take a function as a param as well.

File: src/components/OptionsList.js
Comment (lucas-neuhaus-dev) [reply src/components/OptionsList.js:41]: I actually wanted to discuss this variable. 
Right now it is not just controlling the focus, I created it and went hard on it, just want to get some stuff working.
Right now it is controlling:
- Whether the arrow keys should be on or off in `OptionsSelector`;
- Whether the option can be focused in `OptionsList`;
- Both the text style and if should show the Pencil and PinCircle icons in `OptionRow`.

Maybe we could name it:
- `disableArrowKeysActions` when in the `OptionsSelector` prop;
- `canFocusOptions` when in the `OptionsList` prop;
- `showAdditionalOptionStates` (copied from your review below) in the `OptionRow`

File: src/pages/home/sidebar/OptionRow.js
Comment (lucas-neuhaus-dev) [reply src/pages/home/sidebar/OptionRow.js:64]: I think it makes more sense when we can set the prop to true on calling it.
In this case, instead of naming it `showAdditionalOptionStates`, we name it `hideAdditionalOptionStates`, and default it to false.

File: src/components/CreateMenu.js
Comment (marcaaron) [reply src/components/CreateMenu.js:42]: >take a function as a param as well, so it can be reused if necessary

Will leave this up to you. It would also be fine to wait until something wants to reuse it.

File: src/components/OptionsList.js
Comment (marcaaron) [reply src/components/OptionsList.js:41]: That plan sounds perfect to me.

Comment (lucas-neuhaus-dev) [reply src/components/OptionsList.js:41]: Decided to take the same approach as changing `showAdditionalOptionState` to `hideAdditionalOptionState`, and changed `canFocusOptions` to `disableFocusOptions`.

Also separated it into 2 props on `<OptionsSelector />`. Because there is no correlation between `disableArrowKeysActions` and `hideAdditionalOptionStates`, both should come from this parent.

File: src/components/CreateMenu.js
Comment (lucas-neuhaus-dev) [reply src/components/CreateMenu.js]: Left it as it is.

File: src/components/RightDockedModal.js
Comment (lucas-neuhaus-dev) [reply src/components/RightDockedModal.js]: Done!

File: src/pages/NewGroupPage.js
Comment (lucas-neuhaus-dev) [reply src/pages/NewGroupPage.js]: Done!

File: src/pages/home/sidebar/OptionRow.js
Comment (lucas-neuhaus-dev) [reply src/pages/home/sidebar/OptionRow.js]: Done!

File: src/styles/getModalStyles.js
Comment (lucas-neuhaus-dev) [reply src/styles/getModalStyles.js:126]: Done!

File: src/components/CreateMenu.js
Comment (lucas-neuhaus-dev) [reply src/components/CreateMenu.js:110]: Done!

**Comment by lucas-neuhaus-dev** [comment]
Fixed the code reviews and merged the master branch.

I'm going to take a look at this small width screen issue now.

**Comment by lucas-neuhaus-dev** [comment]
Fixed the problem by calling `hideSidebar()` on small screens when the `Create Group` button is pressed.

Also, there was an error when pressing it with 0 rows selected. Temporarily fixed that by just returning the function if the `array.length < 1`.
It probably needs a better user notification (maybe a toast?) for this case @shawnborton.

File: src/pages/NewGroupPage.js
Comment (marcaaron) [inline src/pages/NewGroupPage.js:155]: Yeah this is fine for now IMO. 
I think ideally we would want to disable the Button until there are at least one selected option.

cc @shawnborton

Comment (shawnborton) [reply src/pages/NewGroupPage.js:155]: I agree with that. I don't know that we have disabled styles handy, but we could make them.

**Review by marcaaron — APPROVED** [review]
LGTM! And thanks for all the changes. Let's proceed 🚀

**Comment by marcaaron** [comment]
One last question I have here that is not a blocker is that all option text appears bold now. But if it's a problem we can address it later.

File: src/pages/NewGroupPage.js
Comment (marcaaron) [reply src/pages/NewGroupPage.js:155]: We should have one for the submit button IIRC

File: src/components/OptionsList.js
Comment (Julesssss) [reply src/components/OptionsList.js]: Hey, does anyone recall why we desire the disabling of the arrow keys for the right docked modal? An issue was opened for enabling the arrow keys, but I can only assume we had a good reason for blocking this in the first place?

Comment (marcaaron) [reply src/components/OptionsList.js]: Sorry, I don't recall why.

**Pull request merged**
